engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/hydration/NadelHydrationInputBuilder.kt

```diff
@@ -1,172 +1,159 @@
-            is NadelHydrationStrategy.OneToOne -> listOf(argsForAllCalls)
-            is NadelHydrationStrategy.ManyToOne -> {
-                val inputDefToSplit = hydrationStrategy.inputDefToSplit
-                val valuesToSplit = when (val valueSource = inputDefToSplit.valueSource) {
-                    is ValueSource.FieldResultValue -> getResultValues(valueSource, parentNode, aliasHelper)
-                    else -> error("Can only split field result value into multiple hydration calls")
-                }
-                valuesToSplit
-                    .asSequence()
-                    .flatten(recursively = true) // Honestly: I think we need to revisit this, we kind of make big assumptions on the schema
-                    .map { value ->
-                        val inputValuePerCall = inputDefToSplit.name to makeNormalizedInputValue(
-                            inputDefToSplit.actorArgumentDef,
-                            value = valueToAstValue(value),
-                        )
-                        argsForAllCalls + inputValuePerCall
+    private fun makeManyToOneArgs(
+        hydrationStrategy: NadelHydrationStrategy.ManyToOne,
+    ): List<Map<String, NormalizedInputValue>> {
+        val inputDefToSplit = hydrationStrategy.inputDefToSplit
+        val valueSourceToSplit = inputDefToSplit.valueSource as ValueSource.FieldResultValue
+        val sharedArgs = makeInputMap(inputDefToSplit)
+
+        return getResultNodes(valueSourceToSplit)
+            .asSequence()
+            .flatMap { node ->
+                // This code belongs together for cohesiveness, do NOT merge it back into the outer sequence
+                sequenceOf(node.value)
+                    .flatten(recursively = true)
+                    .map {
+                        makeInputValue(inputDef = inputDefToSplit, value = it)
                     }
-                    .toList()
             }
-        }
+            .map {
+                // Make the pair to go along with every input map
+                inputDefToSplit.name to it
+            }
+            .map {
+                sharedArgs + it
+            }
+            .filter {
+                isInputMapValid(it)
+            }
+            .toList()
+    }
+
+    private fun makeInputMap(
+        excluding: NadelHydrationActorInputDef? = null,
+    ): Map<String, NormalizedInputValue> {
+        return instruction.actorInputValueDefs
+            .asSequence()
+            .filter {
+                it != excluding
+            }
+            .mapNotNull { inputDef ->
+                makeInputValuePair(inputDef)
+            }
+            .toMapStrictly()
+    }
+
+    /**
+     * Valid in the sense that we should actually query for data.
+     *
+     * If all field values are null, then it doesn't makes sense to actually send the query.
+     */
+    private fun isInputMapValid(inputMap: Map<String, NormalizedInputValue>): Boolean {
```

Comment:
So this is the main change, we check if all the field values are null.

If all the values are null, then unless the underlying service is doing some weird thing, we won't be getting back any data at all.

We can always remove this, this is mainly to just appease the comparison algorithm by matching the current engine.